### PR TITLE
[Merged by Bors] - Fix backward compatibility issues for ManagedConnectorSpec parameters

### LINKED ISSUE #2433: Add custom serializer and deserializer for ManagedConnectorSpec
In #2426, we changed `parameters` from `BTreeMap<String, String>` to `BTreeMap<String, Vec<String>>`:
https://github.com/infinyon/fluvio/blob/7cf1707d70aac4c34de5c892a627c947f8024243/crates/fluvio-controlplane-metadata/src/connector/spec.rs#L10-L27

Let's write a serializer/deserializer to support backward compatibility.

### PR DESCRIPTION
**Pull request by simlay**
Closes #2433.

### CODE REVIEW

File: crates/fluvio-controlplane-metadata/src/connector/spec.rs [hunk 1]
```diff
@@ -22,10 +22,49 @@ pub struct ManagedConnectorSpec {
...
 
     pub topic: String,
-    pub parameters: BTreeMap<String, String>,
+
+    #[fluvio(max_version = 7)]
+    pub parameters_old: BTreeMap<String, String>,
+
+    #[fluvio(min_version = 8)]
+    pub parameters: BTreeMap<String, VecOrString>,
+
```
Comment (simlay) [inline crates/fluvio-controlplane-metadata/src/connector/spec.rs:31]: @morenol is this what you were thinking? I'm not sure I've done it correctly.

File: crates/fluvio-cli/src/connector/mod.rs [hunk 2]
```diff
@@ -165,23 +165,35 @@ impl From<ConnectorConfig> for ManagedConnectorSpec {
...
+                    "producer-linger".to_string(),
+                    ManageConnectorParameterValue::from(vec![linger]),
+                );
             }
             if let Some(compression) = producer.compression {
                 let compression = format!("{:?}", compression);
-                parameters.insert("producer-compression".to_string(), compression);
+                parameters.insert(
+                    "producer-compression".to_string(),
+                    ManageConnectorParameterValue::from(vec![compression]),
```
Comment (morenol) [inline crates/fluvio-cli/src/connector/mod.rs:177]: I think that thid don't need the `vec!`
```suggestion
                    ManageConnectorParameterValue::from(compression),
```

File: crates/fluvio-cli/src/connector/mod.rs [hunk 3]
```diff
@@ -165,23 +165,35 @@ impl From<ConnectorConfig> for ManagedConnectorSpec {
...
+                );
             }
             if let Some(batch_size) = producer.batch_size {
                 let batch_size = format!("{}", batch_size);
-                parameters.insert("producer-batch-size".to_string(), batch_size);
+                parameters.insert(
+                    "producer-batch-size".to_string(),
+                    ManageConnectorParameterValue::from(vec![batch_size]),
```
Comment (morenol) [inline crates/fluvio-cli/src/connector/mod.rs:184]: I think that thid don't need the vec!

File: crates/fluvio-cli/src/connector/mod.rs [hunk 4]
```diff
@@ -165,23 +165,35 @@ impl From<ConnectorConfig> for ManagedConnectorSpec {
         if let Some(producer) = config.producer {
             if let Some(linger) = producer.linger {
                 let linger = humantime::format_duration(linger).to_string();
-                parameters.insert("producer-linger".to_string(), linger);
+                parameters.insert(
+                    "producer-linger".to_string(),
+                    ManageConnectorParameterValue::from(vec![linger]),
```
Comment (morenol) [inline crates/fluvio-cli/src/connector/mod.rs:170]: I think that this don't need the vec!

File: crates/fluvio-cli/src/connector/mod.rs [hunk 5]
```diff
@@ -165,23 +165,35 @@ impl From<ConnectorConfig> for ManagedConnectorSpec {
...
         }
 
         // Consumer arguments are prefixed with `consumer`.
         if let Some(consumer) = config.consumer {
             if let Some(partition) = consumer.partition {
                 let partition = format!("{}", partition);
-                parameters.insert("consumer-partition".to_string(), partition);
+                parameters.insert(
+                    "consumer-partition".to_string(),
+                    ManageConnectorParameterValue::from(vec![partition]),
```
Comment (morenol) [inline crates/fluvio-cli/src/connector/mod.rs:195]: I think that this don't need the vec

File: crates/fluvio-sc-schema/src/objects/list.rs [hunk 6]
```diff
@@ -27,7 +27,7 @@ where
 
 impl Request for ObjectApiListRequest {
     const API_KEY: u16 = AdminPublicApiKey::List as u16;
-    const DEFAULT_API_VERSION: i16 = 6;
+    const DEFAULT_API_VERSION: i16 = 8;
```
Comment (morenol) [inline crates/fluvio-sc-schema/src/objects/list.rs:30]: Probably we need to do this also in `watch.rs`

File: crates/fluvio-sc/src/k8/controllers/managed_connector_deployment.rs [hunk 7]
```diff
@@ -225,7 +228,30 @@ impl ManagedConnectorDeploymentController {
...
+                    for value in values.iter() {
+                        args.push(format!("--{}={}", key.replace('_', "-"), value))
+                    }
+                    args
+                }
+                ManageConnectorParameterValueInner::Map(map) => {
+                    let mut args = Vec::new();
+                    for (sub_key, value) in map.iter() {
+                        args.push(format!(
+                            "--{}-{}={}",
```
Comment (morenol) [inline crates/fluvio-sc/src/k8/controllers/managed_connector_deployment.rs:246]: Do we want this or `--{}="{}={}"`?

Comment (morenol) [reply crates/fluvio-sc/src/k8/controllers/managed_connector_deployment.rs:246]: Not sure for 



```yaml
parameters:
  kafka-options:
    linger.ms: 1
```

Do we want --kafka-options=linger.ms:1 or --kafka-option-linger.ms=1

File: crates/fluvio-cli/src/connector/mod.rs [hunk 8]
```diff
@@ -165,23 +165,35 @@ impl From<ConnectorConfig> for ManagedConnectorSpec {
         if let Some(producer) = config.producer {
             if let Some(linger) = producer.linger {
                 let linger = humantime::format_duration(linger).to_string();
-                parameters.insert("producer-linger".to_string(), linger);
+                parameters.insert(
+                    "producer-linger".to_string(),
+                    ManageConnectorParameterValue::from(linger),
```
Comment (morenol) [inline crates/fluvio-cli/src/connector/mod.rs:170]: nit: Maybe linger.into() is more cleaner

File: crates/fluvio-sc/src/k8/controllers/managed_connector_deployment.rs (hunk 7 above)
Comment (simlay) [reply crates/fluvio-sc/src/k8/controllers/managed_connector_deployment.rs:246]: I think we want `--kafka-options=linger.ms:1 `

File: crates/fluvio-controlplane-metadata/src/connector/spec.rs [hunk 9]
```diff
@@ -22,10 +25,234 @@ pub struct ManagedConnectorSpec {
...
+        {
+            deserializer.deserialize_any(ParameterValueVisitor)
+        }
+    }
+
+    impl<'de> Visitor<'de> for ParameterValueVisitor {
+        type Value = ManageConnectorParameterValueInner;
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("string or map")
```
Comment (nacardin) [inline crates/fluvio-controlplane-metadata/src/connector/spec.rs:73]: string, map, or sequence

File: crates/fluvio-sc/src/k8/controllers/managed_connector_deployment.rs (hunk 7 above)
Comment (simlay) [reply crates/fluvio-sc/src/k8/controllers/managed_connector_deployment.rs:246]: oh, I think I chose wrong.

File: crates/fluvio-cli/src/connector/mod.rs (hunk 8 above)
Comment (simlay) [reply crates/fluvio-cli/src/connector/mod.rs:170]: Yeah it is.

File: crates/fluvio-sc-schema/src/objects/watch.rs [hunk 10]
```diff
@@ -26,7 +26,7 @@ pub struct WatchRequest<S: AdminSpec> {
 
 impl Request for ObjectApiWatchRequest {
     const API_KEY: u16 = AdminPublicApiKey::Watch as u16;
-    const DEFAULT_API_VERSION: i16 = 6;
+    const DEFAULT_API_VERSION: i16 = 8;
```
Comment (simlay) [inline crates/fluvio-sc-schema/src/objects/watch.rs:29]: I don't love this but don't know what it should be. Do we have a global const for these?

Comment (sehz) [reply crates/fluvio-sc-schema/src/objects/watch.rs:29]: They are independent versions so they should not be global.

**Review by nacardin — APPROVED** [review]

**Review by sehz — APPROVED** [review]
LGTM

**Pull request closed without merging**